File: src/main/java/by/home/acs/language/archive/wad/WADArchiveFileSystem.java
package by.home.acs.language.archive.wad;


import by.home.acs.language.archive.BaseArchiveFileSystem;
import com.intellij.openapi.fileTypes.FileType;
import com.intellij.openapi.fileTypes.FileTypeRegistry;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.impl.ArchiveHandler;
import com.intellij.openapi.vfs.newvfs.VfsImplUtil;
import org.jetbrains.annotations.NotNull;

public abstract class WADArchiveFileSystem extends BaseArchiveFileSystem {

    //here was companion object that was working with SevenZip class. But I will skip this and try to see what will be

    @Override
    protected @NotNull ArchiveHandler getHandler(@NotNull VirtualFile entryFile) {
        return VfsImplUtil.getHandler(this, entryFile, WADArchiveHandler::new);
    }

    public ArchiveHandler getHandlerForFile(VirtualFile virtualFile) {
        return getHandler(virtualFile);
    }

    @Override
    protected boolean isCorrectFileType(@NotNull VirtualFile local) {
        return FileTypeRegistry.getInstance().getFileTypeByFileName(local.getName()) instanceof WADFile;
    }
}

File: src/main/kotlin/by/home/acs/language/archive/wad/WADArchiveFileSystem.kt
package by.home.acs.language.archive.wad

import by.home.acs.language.archive.BaseArchiveFileSystem
import by.home.acs.language.archive.util.FileStructureUtil
import com.intellij.openapi.fileTypes.FileType
import com.intellij.openapi.fileTypes.FileTypeRegistry
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.vfs.impl.ArchiveHandler
import com.intellij.openapi.vfs.newvfs.VfsImplUtil
import net.sf.sevenzipjbinding.SevenZip

abstract class WADArchiveFileSystem(
    fileType: FileType,
    fsProtocol: String,
    fsSeparator: String
) : BaseArchiveFileSystem(fsProtocol, fsSeparator) {
    companion object {
        init {
            if (!SevenZip.isInitializedSuccessfully()) {
                val tmpDir = FileStructureUtil.getPluginTempFolder()
                if (!tmpDir.exists()) {
                    tmpDir.mkdirs()
                }
                SevenZip.initSevenZipFromPlatformJAR(tmpDir)
            }
        }
    }

    override fun getHandler(entryFile: VirtualFile): ArchiveHandler {
        return VfsImplUtil.getHandler(this, entryFile) { localPath ->
            WADArchiveHandler(localPath)
        }
    }

    fun getHandlerForFile(file: VirtualFile): ArchiveHandler? {
        return getHandler(file)
    }

    override fun isCorrectFileType(local: VirtualFile): Boolean {
        return FileTypeRegistry.getInstance().getFileTypeByFileName(local.name) is WADArchiveFileType
    }
}